src/v/serde/serde.h

```diff
@@ -171,53 +170,69 @@ void write(iobuf& out, T t) {
 }
 
 template<typename T>
-std::decay_t<T> read(iobuf_parser&);
+std::decay_t<T> read(iobuf_parser&, std::size_t bytes_left_limit);
 
 template<typename T>
-header_t read_header(iobuf_parser& in) {
+std::decay_t<T> read(iobuf_parser& in) {
+    return read<T>(in, 0U);
+}
+
+template<typename T>
+header read_header(iobuf_parser& in, std::size_t const bytes_left_limit) {
     using Type = std::decay_t<T>;
 
-    auto const version = read<version_t>(in);
-    auto const compat_version = read<version_t>(in);
-    auto const size = read<serde_size_t>(in);
+    auto const version = read<version_t>(in, bytes_left_limit);
+    auto const compat_version = read<version_t>(in, bytes_left_limit);
+    auto const size = read<serde_size_t>(in, bytes_left_limit);
 
     if (unlikely(compat_version > Type::redpanda_serde_version)) {
         throw serde_exception(fmt_with_ctx(
           ssx::sformat,
-          "read compat_version={} > {}::version={}",
+          "read {}: compat_version={} > {}::version={}",
+          type_str<Type>(),
           static_cast<int>(compat_version),
           type_str<T>(),
           static_cast<int>(Type::redpanda_serde_version)));
     }
 
-    if (unlikely(in.bytes_left() < size)) {
-        throw serde_exception(fmt_with_ctx(
-          ssx::sformat,
-          "bytes_left={}, size={}",
-          in.bytes_left(),
-          static_cast<int>(size)));
-    }
```

Comment:
In general if we have a container of bytes from a newer version there might have been some extra fields, in which case, an older client that receives the container could still decode the data corresponding to the older version and there would be bytes left, which would be valid.

But... in this exact case, it doesn't have anything to do with the size of the type at our version but rather its just internally in the buffer we have something that doesn't match so the bytes left don't match the communicated size so there is something strange going on.

Todo items

1. let's check this exact case
2. should we make a strict mode?
3. add units tests for these cases

Reply:
It's safe to merge with this, and we'll add in these additional checks later.

Reply:
Check added again. It's useful to catch errors.